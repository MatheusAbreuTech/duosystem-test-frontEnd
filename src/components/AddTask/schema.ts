import { z } from 'zod'

export const schema = z.object({
  taskDescription: z.string().min(1, 'Informe uma descrição para a tarefa.')
})
