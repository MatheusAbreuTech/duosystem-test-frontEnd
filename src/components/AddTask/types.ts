import { z } from 'zod'
import { schema } from './schema'

export type SchemaType = z.infer<typeof schema>
