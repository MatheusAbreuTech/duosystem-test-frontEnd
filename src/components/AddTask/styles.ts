import styled from 'styled-components'

export const Form = styled.form`
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  width: 100%;
  align-items: flex-start;
  gap: 0.5rem;
`

export const InputWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
  width: 100%;
`
