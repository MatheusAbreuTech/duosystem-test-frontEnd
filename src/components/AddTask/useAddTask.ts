import useTaskStore from '@/store/zustand'
import { SubmitHandler, useForm } from 'react-hook-form'
import { SchemaType } from './types'
import { schema } from './schema'
import { zodResolver } from '@hookform/resolvers/zod'

export const useAddTask = () => {
  const addTask = useTaskStore((state) => state.addTask)
  const methods = useForm<SchemaType>({
    resolver: zodResolver(schema)
  })

  const {
    handleSubmit,
    reset,
    formState: { errors }
  } = methods

  const handleAddTask: SubmitHandler<SchemaType> = (data) => {
    const { taskDescription } = data

    addTask(taskDescription)

    reset()
  }

  return {
    errors,
    methods,
    handleSubmit,
    handleAddTask
  }
}
