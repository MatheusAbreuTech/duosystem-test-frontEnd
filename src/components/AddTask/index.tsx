'use client'

import * as S from './styles'

import { FormProvider } from 'react-hook-form'
import Input from '../Input'
import Button from '../Button'
import ErrorMessage from '../ErrorMessage'
import { useAddTask } from './useAddTask'

const AddTask = () => {
  const { methods, errors, handleSubmit, handleAddTask } = useAddTask()

  return (
    <FormProvider {...methods}>
      <S.Form onSubmit={handleSubmit(handleAddTask)}>
        <label htmlFor="taskDescription">Adicionar tarefa:</label>

        <S.InputWrapper>
          <Input
            type="text"
            id="taskDescription"
            name="taskDescription"
            data-testid="taskDescription"
          />
          <Button color="darkBlue" type="submit">
            Adicionar
          </Button>
        </S.InputWrapper>
      </S.Form>
      <ErrorMessage errors={errors} field="taskDescription" />
    </FormProvider>
  )
}

export default AddTask
