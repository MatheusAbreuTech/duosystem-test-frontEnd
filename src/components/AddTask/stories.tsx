import { Meta, StoryObj } from '@storybook/react'
import AddTask from '.'

export default {
  title: 'AddTask',
  component: AddTask
} as Meta

export const Default: StoryObj = {}
