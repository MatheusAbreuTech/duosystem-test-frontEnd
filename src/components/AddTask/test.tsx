import { fireEvent, render, waitFor } from '@testing-library/react'

import AddTask from '.'

describe('<AddTask />', () => {
  it('should render the AddTask component', () => {
    const { container } = render(<AddTask />)

    expect(container).toBeInTheDocument()
  })

  it('should call the submit function and clear the text field', async () => {
    const { getByLabelText, getByText } = render(<AddTask />)

    const input = getByLabelText('Adicionar tarefa:')

    fireEvent.change(input, { target: { value: 'Nova Tarefa' } })

    const addButton = getByText('Adicionar')

    fireEvent.click(addButton)

    await waitFor(() => {
      expect(input).toHaveValue('')
    })
  })
})
