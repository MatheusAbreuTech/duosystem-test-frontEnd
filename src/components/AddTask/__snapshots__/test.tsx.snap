// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTask /> should render the AddTask component 1`] = `
.c0 {
  display: flex;
  justify-content: space-between;
  flex-direction: column;
  width: 100%;
  align-items: flex-start;
  gap: 0.5rem;
}

.c0 >label {
  font-size: 1.6rem;
}

.c1 {
  display: flex;
  align-items: center;
  gap: 1rem;
  width: 100%;
}

.c1 >input {
  border: 1px solid #383838;
}

.c2 {
  width: 100%;
  height: 42px;
  padding: 0 1rem;
  background-color: #fff;
  outline: none;
  border: none;
  font-size: 1.6rem;
  border-radius: 5px;
}

.c3 {
  background-color: #0c2e73;
  color: #ffffff;
  border: none;
  border-radius: 0.5rem;
  padding: 1rem;
  max-height: 4.2rem;
  height: 100%;
  cursor: pointer;
  font-size: 1.6rem;
}

<form
  class="c0"
>
  <label
    for="taskDescription"
  >
    Adicionar tarefa:
  </label>
  <div
    class="c1"
  >
    <input
      class="c2"
      data-testid="taskDescription"
      id="taskDescription"
      name="taskDescription"
      type="text"
    />
    <button
      class="c3"
      color="darkBlue"
      type="submit"
    >
      Adicionar
    </button>
  </div>
</form>
`;
